### PR TITLE
Core: Support filtering by regular expression

### PR DESCRIPTION
**Pull request by trentmwillis**
Introduces a "regexFilter" config option that, when enabled, treats the "filter"
value as a regular expression to test against instead of a string.

Fixes: #904

### CODE REVIEW

**Comment by leobalter** [comment]
It's great and includes proper tests for it, even though, I believe controlling regex via config/url-param is not greatly discoverable or might be confusing when defined on the code config.

I'm not silver on this, but I still prefer the approach of triggering the regexFilter if the input field has a valid regexp literal notation.

@gibson042 @jzaefferer @rwjblue: would you give some feedback here? I believe this deserves a discussion before shipping.

**Comment by gibson042** [comment]
screen

I really like the tests, and how clean the code is. However, I don't like the [toolbar clutter](https://jsfiddle.net/zos5whbh/)—especially with the separation between the new checkbox and what it affects—and the forced lowercasing not only precludes case-sensitive regular expression filters, but actually introduces bugs (since `\B`, `\D`, `\S`, and `\W` get converted into their negations).

Since we've already set the precedent for in-band signaling with negation-by-leading-`!`, I'm in favor of continuing that by triggering on regex literal syntax like @leobalter suggested (e.g., `isRegex = /^!?\/[\w\W]*\/i?$/.test( filter )`). An alternative would be a `filterType` parameter and one button for each value, but that just doesn't seem as user-friendly.

**Comment by trentmwillis** [comment]
Turning off the forced lowercasing wouldn't be difficult (didn't think to do it before) and will need to happen even if we move to an in-band approach.

I'm not totally opposed to triggering on regex literals, but as I mentioned before that does become a slightly backwards incompatible change (e.g., filtering for `/components/` will now behave differently) and can result in some confusing functionality (e.g., filtering for `/components/i` is treated as a regex, but `/components/a` is not). If those drawbacks are acceptable I can try implementing the other solution.

Another option would be to restructure the toolbar such that toggling regex behavior is associated better with the filter, but that will involve some UX problems to solve.

**Comment by trentmwillis** [comment]
@leobalter @gibson042 after hearing from @jzaefferer that he would prefer not adding a new config option as well, I've updated to check for regex literals.

File: src/test.js [hunk 1]
```diff
@@ -356,8 +356,8 @@ Test.prototype = {
 	},
 
 	valid: function() {
-		var include,
-			filter = config.filter && config.filter.toLowerCase(),
+		var filter = config.filter,
+			regexFilter = /^!?\/([\w\W]*)\/(i?$)/.exec( filter ),
```
Comment (leobalter) [inline src/test.js]: this regexp also capture strings starting with `!`. IMHO, we may remove it and go with `/^\/([\w\W]*)\/(i?$)/` only.

The `!` prefix flag is good for is strings but not necessary at all for regexp.

Also the current regexFilter function is not handling this prefix.

Removing the `!` flag for regexps is not a breaking feature as it is still operational for regular string searches.

File: test/regex-filter.html [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    <script src="../dist/qunit.js"></script>
+    <script>
+    QUnit.config.filter = '/foo|bar/i'
+    </script>
+    <script src="regex-filter.js"></script>
+  </head>
+  <body>
+    <div id="qunit"></div>
+  </body>
+</html>
```
Comment (leobalter) [inline test/regex-filter.html]: Would you change the indentation from space to tabs? The same goes to:

- test/regex-filter.js
- test/string-filter.html
- test/string-filter.js

**Comment by leobalter** [comment]
It's looking good, with only minor changes to go.

File: src/test.js (hunk 1 above)
Comment (trentmwillis) [inline src/test.js]: Forgot to include that in the functionality. Personally, I think it would be a nice convenience feature since writing converse regexs can get ugly. I don't see a big downside to including it, but there are some benefits.

Comment (leobalter) [inline src/test.js]: Seems reasonable. If we include it on the method and add tests for it, I'm game.

File: src/test.js [hunk 3]
```diff
@@ -356,8 +356,8 @@ Test.prototype = {
...
 			module = QUnit.urlParams.module && QUnit.urlParams.module.toLowerCase(),
 			fullName = ( this.module.name + ": " + this.testName ).toLowerCase();
```
Comment (gibson042) [inline src/test.js]: I think this `.toLowerCase()` should move into `stringFilter`.

Comment (trentmwillis) [inline src/test.js]: Yeah, caught that too when writing more tests.

**Comment by trentmwillis** [comment]
Updated with support for excluding regex matches and additional tests. Also fixed spacing issues.

File: src/test.js [hunk 4]
```diff
@@ -389,7 +389,23 @@ Test.prototype = {
...
+		return regexFilter ?
+			this.regexFilter( regexFilter[1], regexFilter[2], regexFilter[3], fullName ) :
+			this.stringFilter( filter, fullName );
+	},
+
+	regexFilter: function( exclude, pattern, flags, fullName ) {
+		var regex = new RegExp( pattern, flags );
+		var include = regex.test( fullName );
+
+		return exclude ? !include : include;
```
Comment (gibson042) [inline src/test.js]: This `include` variable is dependent upon `fullName`, unlike the `include` variable in `stringFilter`. But honestly, I don't think it improves clarity anyway. What about something like
```js
var match = regex.test( fullName );
return match === !exclude;
```
?

`stringFilter` can be similarly cleaned up in another PR.

**Comment by gibson042** [comment]
One last nit, but LGTM!

File: src/test.js (hunk 4 above)
Comment (trentmwillis) [inline src/test.js]: Makes sense. I went back and forth on what to name this. Will update.

**Comment by trentmwillis** [comment]
@gibson042 @leobalter PR has been updated with the last minor changes.

**Comment by trentmwillis** [comment]
@leobalter any chance this can land soon?

**Comment by leobalter** [comment]
I'm on vacation this week (from mobile only, no computers around). @gibson042 might be able to help, if not traveling for the holidays as well.

I'll be back on Sunday.

**Comment by leobalter** [comment]
Merged at e972135.

Thanks, @trentmwillis!
